File: src/core/files.js
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { existsSync, readdirSync, statSync } from "node:fs";
import { dirname, join, relative } from "node:path";

const DEFAULT_IGNORES = new Set([
  ".git",
  "node_modules",
  "vendor",
  "dist",
  "build",
  "coverage",
  ".cache",
  "runs",
  "workspaces"
]);

export async function readText(path) {
  return await readFile(path, "utf8");
}

export async function writeJson(path, value) {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, JSON.stringify(value, null, 2));
}

export async function writeText(path, value) {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, value);
}

export function walkFiles(root, options = {}) {
  const extensions = options.extensions ?? null;
  const ignores = new Set([...(options.ignores ?? []), ...DEFAULT_IGNORES]);
  const files = [];

  function walk(dir) {
    for (const entry of readdirSync(dir)) {
      if (ignores.has(entry)) {
        continue;
      }
      const full = join(dir, entry);
      const stat = statSync(full);
      if (stat.isDirectory()) {
        walk(full);
      } else if (!extensions || extensions.some((ext) => entry.endsWith(ext))) {
        files.push({
          path: full,
          relativePath: relative(root, full),
          size: stat.size
        });
      }
    }
  }

  if (existsSync(root)) {
    walk(root);
  }
  return files;
}

File: go_issue_agent/core/files.py
from __future__ import annotations

import json
from pathlib import Path


DEFAULT_IGNORES = {
    ".git",
    "node_modules",
    "vendor",
    "dist",
    "build",
    "coverage",
    ".cache",
    "runs",
    "workspaces",
    "__pycache__",
}


def read_text(path: str | Path) -> str:
    return Path(path).read_text(encoding="utf-8")


def write_text(path: str | Path, value: str) -> None:
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(value, encoding="utf-8")


def write_json(path: str | Path, value) -> None:
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(json.dumps(value, indent=2), encoding="utf-8")


def walk_files(root: str | Path, extensions: list[str] | None = None, ignores: list[str] | None = None):
    root_path = Path(root)
    if not root_path.exists():
        return []

    ignore_set = set(DEFAULT_IGNORES)
    if ignores:
        ignore_set.update(ignores)

    files = []
    for path in root_path.rglob("*"):
        if any(part in ignore_set for part in path.parts[len(root_path.parts) :]):
            continue
        if path.is_symlink() or not path.is_file():
            continue
        if extensions and path.suffix not in extensions:
            continue
        files.append(
            {
                "path": str(path),
                "relativePath": str(path.relative_to(root_path)),
                "size": path.stat().st_size,
            }
        )
    return files
